File: solutions/ruby/perfect-numbers/1/perfect_numbers.rb
module PerfectNumber
  def self.classify(number)
    raise ArgumentError, 'Classification is only possible for positive integers.' unless number.positive?

    a = (1..number / 2).filter { (number % _1).zero? }.sum
    if a < number then 'deficient'
    elsif a > number then 'abundant'
    else 'perfect'
    end
  end
end

File: solutions/clojure/perfect-numbers/1/src/perfect_numbers.clj
(ns perfect-numbers)

(defn- aliquot-sum [n]
  (if (= n 1) 0
      (->> (range 2 (inc (long (Math/sqrt n))))
           (filter #(zero? (mod n %)))
           (mapcat (juxt identity #(quot n %)))
           distinct
           (reduce + 1))))

(defn classify [n]
  (case (compare (aliquot-sum n) n)
    -1 :deficient
    0 :perfect
    1 :abundant))
